
This reverts 2f0810880f082fa8ba66ab2c33b02e4ff9770a5e which breaks changing
metadata profile rebalancing.

See: http://www.spinics.net/lists/linux-btrfs/msg42084.html

--- b/fs/btrfs/extent-tree.c
+++ a/fs/btrfs/extent-tree.c
@@ -8482,6 +8482,14 @@
 	if (IS_ERR(trans))
 		return PTR_ERR(trans);
 
+	alloc_flags = update_block_group_flags(root, cache->flags);
+	if (alloc_flags != cache->flags) {
+		ret = do_chunk_alloc(trans, root, alloc_flags,
+				     CHUNK_ALLOC_FORCE);
+		if (ret < 0)
+			goto out;
+	}
+
 	ret = set_block_group_ro(cache, 0);
 	if (!ret)
 		goto out;
@@ -8492,11 +8500,6 @@
 		goto out;
 	ret = set_block_group_ro(cache, 0);
 out:
-	if (cache->flags & BTRFS_BLOCK_GROUP_SYSTEM) {
-		alloc_flags = update_block_group_flags(root, cache->flags);
-		check_system_chunk(trans, root, alloc_flags);
-	}
-
 	btrfs_end_transaction(trans, root);
 	return ret;
 }
